bolt-boost/src/server.rs

```diff
@@ -0,0 +1,582 @@
+    let max_timeout_ms = state.pbs_config().timeout_get_header_ms.min(
+        state
+            .pbs_config()
+            .late_in_slot_time_ms
+            .saturating_sub(ms_into_slot),
+    );
+
+    if max_timeout_ms == 0 {
+        warn!(
+            ms_into_slot,
+            threshold = state.pbs_config().late_in_slot_time_ms,
+            "late in slot, skipping relay requests"
+        );
+
+        return Ok(StatusCode::NO_CONTENT.into_response());
+    }
+
+    let (_, slot_uuid) = state.get_slot_and_uuid();
+
+    // prepare headers, except for start time which is set in `send_one_get_header`
+    let mut send_headers = HeaderMap::new();
+    // TODO: error handling
+    send_headers.insert(
+        HEADER_SLOT_UUID_KEY,
+        HeaderValue::from_str(&slot_uuid.to_string()).unwrap(),
+    );
+    send_headers.insert(
+        USER_AGENT,
+        get_user_agent_with_version(&req_headers).unwrap(),
+    );
+
+    let relays = state.relays();
+    let mut handles = Vec::with_capacity(relays.len());
+    for relay in relays.iter() {
+        handles.push(send_timed_get_header(
+            params,
+            relay.clone(),
+            state.config.chain,
+            state.pbs_config(),
+            send_headers.clone(),
+            ms_into_slot,
+            max_timeout_ms,
+        ));
+    }
+
+    let results = join_all(handles).await;
+    let mut relay_bids = Vec::with_capacity(relays.len());
+    let mut hash_to_proofs = HashMap::new();
+
+    // Get and remove the constraints for this slot
+    let constraints = state.data.constraints.remove(params.slot);
+
+    for (i, res) in results.into_iter().enumerate() {
+        let relay_id = relays[i].id.as_ref();
+
+        match res {
+            Ok(Some(res)) => {
+                let root = res.data.header.message.header.transactions_root;
+
+                let start = Instant::now();
+
+                // Verify the multiproofs
+                if let Err(e) =
+                    verify_multiproofs(constraints.as_ref().unwrap(), &res.data.proofs, root)
+                {
+                    error!(?e, relay_id, "Failed to verify multiproof, skipping bid");
+                    continue;
+                }
+                tracing::debug!("Verified multiproof in {:?}", start.elapsed());
+
+                // Save the proofs per block hash
+                hash_to_proofs.insert(res.data.header.message.header.block_hash, res.data.proofs);
+
+                let vanilla_response = GetHeaderResponse {
+                    version: res.version,
+                    data: res.data.header,
+                };
+
+                relay_bids.push(vanilla_response)
+            }
+            Ok(_) => {}
+            Err(err) if err.is_timeout() => error!(err = "Timed Out", relay_id),
+            Err(err) => error!(?err, relay_id),
+        }
+    }
+
+    let header = state.add_bids(params.slot, relay_bids);
```

Comment:
it's not intuitive that this function returns the best bid in the current cache. can we make it more explicit by renaming it to `best_header` or `best_relay_bid`?

Reply:
Agreed, this is in commit-boost however